src/OpenCV/Contrib/Juicy.hs

```diff
@@ -0,0 +1,179 @@
+type instance JD PixelCMYK8 = Word8
+type instance JD PixelCMYK16 = Word16
+
+type instance JC Pixel8 = 1
+type instance JC Pixel16 = 1
+type instance JC PixelF = 1
+type instance JC PixelYA8 = 2
+type instance JC PixelYA16 = 2
+type instance JC PixelRGB8 = 3
+type instance JC PixelRGB16 = 3
+type instance JC PixelRGBF = 3
+type instance JC PixelRGBA8 = 4
+type instance JC PixelRGBA16 = 4
+type instance JC PixelYCbCr8 = 3
+type instance JC PixelCMYK8 = 4
+type instance JC PixelCMYK16 = 4
+
+-- | An OpenCV bidimensional matrix
+type Mat2D h w c d = Mat ('S '[h,w]) c d 
+
+-- | Compute an OpenCV 2D-matrix from a JuicyPixels image
+fromImage 
+    :: forall a c d . (ToDepth (Proxy d), KnownNat c, Pixel a, Storable a, c ~ JC a, d ~ JD a) 
+    => Image a -- ^ JuicyPixels image
+    -> CvExcept (Mat2D 'D 'D ('S c) ('S d)) 
+fromImage i@(Image h w v) = withMatM 
+    (fi h ::: fi w ::: Z) 
+    (Proxy :: Proxy c) 
+    (Proxy :: Proxy d) 
+    (pure 0 :: V4 Double) $ \m -> 
+        forM_ ((,) <$> [0 .. h - 1] <*> [0 .. w - 1]) $ \(x,y) -> unsafeWrite m [y,x] (pixelAt i x y) 
+  where
+    fi :: Int -> Int32
+    fi = fromIntegral 
+
+-- | Compute a JuicyPixels image from an OpenCV 2D-matrix
+toImage :: forall a c d . (KnownNat c, Pixel a, Storable a, c ~ JC a, d ~ JD a) 
+    => Mat2D 'D 'D ('S c) ('S d)  -- ^ OpenCV 2D-matrix
+    -> Image a
+toImage m  = let
+    MatInfo [fromIntegral -> h, fromIntegral -> w] _ _  = matInfo m
+    in unsafePerformIO $ do
+        mat <- unsafeThaw m
+        withImage h w $ \x y -> unsafeRead mat [y,x] >>= peek
+      
+-- | An OpenCV 2D-filter preserving the matrix type
+type Filter m h w c d = Mat2D h w c d -> CvExceptT m (Mat2D h w c d)
+
+-- | Apply an OpenCV 2D-filter to a JuicyPixels dynamic matrix, preserving the Juicy pixel encoding
+isoJuicy :: forall m c d . (PrimMonad m, ToDepthDS (Proxy d), ToNatDS (Proxy c)) 
+    => Filter m 'D 'D c d -- ^ OpenCV 2D-filter
+    -> DynamicImage -- ^ JuicyPixels dynamic image
+    -> CvExceptT m DynamicImage 
+
+isoJuicy f (ImageRGB8 i)    =  ImageRGB8    <$> isoApply f i 
+isoJuicy f (ImageRGB16 i)   =  ImageRGB16   <$> isoApply f i 
+isoJuicy f (ImageRGBF i)    =  ImageRGBF    <$> isoApply f i 
+isoJuicy f (ImageY8 i)      =  ImageY8      <$> isoApply f i 
+isoJuicy f (ImageY16 i)     =  ImageY16     <$> isoApply f i 
+isoJuicy f (ImageRGBA8 i)   =  ImageRGBA8   <$> isoApply f i 
+isoJuicy f (ImageRGBA16 i)  =  ImageRGBA16  <$> isoApply f i 
+
+-- apply a filter for a specific 'a'. Coerce failures are lifted to the monadic layer
+isoApply ::
+     (PrimMonad m, KnownNat (JC a), Pixel a, Storable a, 
+     ToShapeDS (Proxy ('S '[h,w])), 
+     ToChannelsDS (Proxy c), 
+     ToDepthDS (Proxy d), ToChannelsDS (Proxy ('S (JC a))), 
+     ToDepthDS (Proxy ('S (JD a))),
+     ToDepth (Proxy (JD a))
+     )
+  => Filter m h w c d -> Image a -> CvExceptT m (Image a)
+
+isoApply f i = toImage <$> (pureExcept (fromImage i >>= coerceMat) >>= f >>= pureExcept . coerceMat)
```

Comment:
If we go with the non-failing `fromImage` function we can have the more general and more efficient (because there is no `coerceMat`):

```haskell
isoApply
    :: forall f inPixel outPixel
    . ( Functor f
      , KnownNat (JC inPixel)
      , KnownNat (JC outPixel)
      , Storable inPixel
      , Storable outPixel
      , ToDepth (Proxy (JD inPixel))
      , Pixel inPixel
      , Pixel outPixel
      )
    => (     Mat2D 'D 'D ('S (JC inPixel))  ('S (JD inPixel))
       -> f (Mat2D 'D 'D ('S (JC outPixel)) ('S (JD outPixel)))
       )
    -> (     Image inPixel
       -> f (Image outPixel)
       )
isoApply f i = toImage <$> f (fromImage i)
```

Then the type signature of `isoJuicy` needs to be:

```haskell
isoJuicy :: forall m. (PrimMonad m)
    => (forall c d. Filter m 'D 'D c d) -- ^ OpenCV 2D-filter
    -> DynamicImage -- ^ JuicyPixels dynamic image
    -> CvExceptT m DynamicImage
```

which seems quite limited. Maybe it's better to not provide that function but export `isoApply` instead. @roelvandijk what do you think?